In agent.stream() and agent.astream() yield actions before observations

nfcampos opened the pull request:
- This emits actions as soon as they are available, ie. right after plan
- When action is executed an ActionObservation object is emitted, which contains both the action and the return value of the tool

```
---- chunk 1
{'actions': [AgentActionMessageLog(tool='Search', tool_input="Leo DiCaprio's current girlfriend", log="\nInvoking: `Search` with `Leo DiCaprio's current girlfriend`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'function_call': {'name': 'Search', 'arguments': '{\n  "__arg1": "Leo DiCaprio\'s current girlfriend"\n}'}})])]}
---- chunk 2
{'observations': [AgentObservation(action=AgentActionMessageLog(tool='Search', tool_input="Leo DiCaprio's current girlfriend", log="\nInvoking: `Search` with `Leo DiCaprio's current girlfriend`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'function_call': {'name': 'Search', 'arguments': '{\n  "__arg1": "Leo DiCaprio\'s current girlfriend"\n}'}})]), observation="Vittoria Ceretti's")]}
---- chunk 3
{'actions': [AgentActionMessageLog(tool='Search', tool_input="Vittoria Ceretti's current age", log="\nInvoking: `Search` with `Vittoria Ceretti's current age`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'function_call': {'name': 'Search', 'arguments': '{\n  "__arg1": "Vittoria Ceretti\'s current age"\n}'}})])]}
---- chunk 4
{'observations': [AgentObservation(action=AgentActionMessageLog(tool='Search', tool_input="Vittoria Ceretti's current age", log="\nInvoking: `Search` with `Vittoria Ceretti's current age`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'function_call': {'name': 'Search', 'arguments': '{\n  "__arg1": "Vittoria Ceretti\'s current age"\n}'}})]), observation='25 years')]}
---- chunk 5
{'actions': [AgentActionMessageLog(tool='Calculator', tool_input='25^0.43', log='\nInvoking: `Calculator` with `25^0.43`\n\n\n', message_log=[AIMessageChunk(content='', additional_kwargs={'function_call': {'name': 'Calculator', 'arguments': '{\n  "__arg1": "25^0.43"\n}'}})])]}
---- chunk 6
{'observations': [AgentObservation(action=AgentActionMessageLog(tool='Calculator', tool_input='25^0.43', log='\nInvoking: `Calculator` with `25^0.43`\n\n\n', message_log=[AIMessageChunk(content='', additional_kwargs={'function_call': {'name': 'Calculator', 'arguments': '{\n  "__arg1": "25^0.43"\n}'}})]), observation='Answer: 3.991298452658078')]}
---- chunk 7
{'output': "Leo DiCaprio's current girlfriend is Vittoria Ceretti, and her age "
           '(25 years) raised to the 0.43 power is approximately 3.99.'}
---- final
{'actions': [AgentActionMessageLog(tool='Search', tool_input="Leo DiCaprio's current girlfriend", log="\nInvoking: `Search` with `Leo DiCaprio's current girlfriend`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'function_call': {'name': 'Search', 'arguments': '{\n  "__arg1": "Leo DiCaprio\'s current girlfriend"\n}'}})]),
             AgentActionMessageLog(tool='Search', tool_input="Vittoria Ceretti's current age", log="\nInvoking: `Search` with `Vittoria Ceretti's current age`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'function_call': {'name': 'Search', 'arguments': '{\n  "__arg1": "Vittoria Ceretti\'s current age"\n}'}})]),
             AgentActionMessageLog(tool='Calculator', tool_input='25^0.43', log='\nInvoking: `Calculator` with `25^0.43`\n\n\n', message_log=[AIMessageChunk(content='', additional_kwargs={'function_call': {'name': 'Calculator', 'arguments': '{\n  "__arg1": "25^0.43"\n}'}})])],
 'observations': [AgentObservation(action=AgentActionMessageLog(tool='Search', tool_input="Leo DiCaprio's current girlfriend", log="\nInvoking: `Search` with `Leo DiCaprio's current girlfriend`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'function_call': {'name': 'Search', 'arguments': '{\n  "__arg1": "Leo DiCaprio\'s current girlfriend"\n}'}})]), observation="Vittoria Ceretti's"),
                  AgentObservation(action=AgentActionMessageLog(tool='Search', tool_input="Vittoria Ceretti's current age", log="\nInvoking: `Search` with `Vittoria Ceretti's current age`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'function_call': {'name': 'Search', 'arguments': '{\n  "__arg1": "Vittoria Ceretti\'s current age"\n}'}})]), observation='25 years'),
                  AgentObservation(action=AgentActionMessageLog(tool='Calculator', tool_input='25^0.43', log='\nInvoking: `Calculator` with `25^0.43`\n\n\n', message_log=[AIMessageChunk(content='', additional_kwargs={'function_call': {'name': 'Calculator', 'arguments': '{\n  "__arg1": "25^0.43"\n}'}})]), observation='Answer: 3.991298452658078')],
 'output': "Leo DiCaprio's current girlfriend is Vittoria Ceretti, and her age "
           '(25 years) raised to the 0.43 power is approximately 3.99.'}
```

nfcampos committed:
In agent.stream() and agent.astream() yield actions before observations

- This emits actions as soon as they are available, ie. right after plan
- When action is executed an ActionObservation object is emitted, which contains both the action and the return value of the tool

eyurtsev commented on libs/langchain/langchain/agents/agent.py:
```diff
@@ -925,6 +942,27 @@ def _take_next_step(
     ) -> Union[AgentFinish, List[Tuple[AgentAction, str]]]:
+        return self._consume_next_step(
+            [
+                a
```
a -> next_step?? a better variable name would be useful

hwchase17 commented on libs/langchain/langchain/schema/agent.py:
```diff
@@ -50,6 +50,15 @@ class AgentActionMessageLog(AgentAction):
     type: Literal["AgentActionMessageLog"] = "AgentActionMessageLog"  # type: ignore
 
 
+class AgentObservation(Serializable):
```
AgentObservation.observation seems weird namespacing

maybe AgentTurn? AgentIteration? AgentStep?

nfcampos committed:
Rename to AgentStep

eyurtsev commented on libs/langchain/langchain/agents/agent.py:
```diff
@@ -1026,6 +1066,27 @@ async def _atake_next_step(
+            [
+                a
+                async for a in self._aiter_next_step(
+                    name_to_tool_map,
```
variable name

eyurtsev reviewed (approved):
Looks reasonable just minor nits for variable name

nfcampos committed:
Fix configurable for non-hashable options

nfcampos committed:
Fix type

nfcampos committed:
Add "messages" key to output of AgentExecutor.stream()

Lint

Fix output passed to tracer, fix tests

Lint

nfcampos committed:
different fix

nfcampos committed:
Fix configurable for non-hashable options (#12890)



<!-- Thank you for contributing to LangChain!

Replace this entire comment with:
  - **Description:** a description of the change, 
  - **Issue:** the issue # it fixes (if applicable),
  - **Dependencies:** any dependencies required for this change,
- **Tag maintainer:** for a quicker response, tag the relevant
maintainer (see below),
- **Twitter handle:** we announce bigger features on Twitter. If your PR
gets announced, and you'd like a mention, we'll gladly shout you out!

Please make sure your PR is passing linting and testing before
submitting. Run `make format`, `make lint` and `make test` to check this
locally.

See contribution guidelines for more information on how to write/run
tests, lint, etc:

https://github.com/langchain-ai/langchain/blob/master/.github/CONTRIBUTING.md

If you're adding a new integration, please include:
1. a test for the integration, preferably unit tests that do not rely on
network access,
2. an example notebook showing its use. It lives in `docs/extras`
directory.

If no one reviews your PR within a few days, please @-mention one of
@baskaryan, @eyurtsev, @hwchase17.
 -->
